LUCENE-10216: Use MergeScheduler and MergePolicy to run addIndexes(CodecReader[]) merges.

vigyasharma opened the pull request:
<!--
_(If you are a project committer then you may remove some/all of the following template.)_

Before creating a pull request, please file an issue in the ASF Jira system for Lucene:

* https://issues.apache.org/jira/projects/LUCENE

You will need to create an account in Jira in order to create an issue.

The title of the PR should reference the Jira issue number in the form:

* LUCENE-####: <short description of problem or changes>

LUCENE must be fully capitalized. A short description helps people scanning pull requests for items they can work on.

Properly referencing the issue in the title ensures that Jira is correctly updated with code review comments and commits. -->


# Description

The addIndexes(CodecReader...) API today merges all provided readers into a single merge, in one large blocking call. We want to add concurrency here by invoking multiple parallel merges on subsets of readers, in a way that is configurable by users. The merged segments created, can later be merged further in the regular, non-blocking, background merges triggered by Lucene. Currently, users are responsible  for managing their API run times, by invoking it multiple times with subsets of readers. 
JIRA - https://issues.apache.org/jira/browse/LUCENE-10216


# Solution

In this change, we leverage the configured MergeScheduler to invoke all merges required by the addIndexes API. MergePolicy also exposes a `findMerges(List<CodecReader> readers)` API that users can override to configure how their readers should be clustered into merges. We add a default implementation that retains current behavior of adding all readers into a single merge.


# Tests

- Existing tests passing
- Pending: New tests to be added

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/lucene/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Lucene maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.

vigyasharma committed:
Use merge policy and merge scheduler to run addIndexes merges

vigyasharma committed:
wrapped reader does not see deletes - debug

vigyasharma committed:
Partially fixed tests in TestAddIndexes

vigyasharma committed:
Use writer object to invoke addIndexes merge

vigyasharma committed:
Use merge object info

vigyasharma committed:
Add javadocs for new methods

vigyasharma committed:
TestAddIndexes passing

vigyasharma committed:
verify field info schemas upfront from incoming readers

vigyasharma committed:
rename flag to track pooled readers

vigyasharma committed:
Keep addIndexes API transactional

vigyasharma committed:
Maintain transactionality - register segments with iw after all merges complete

vigyasharma committed:
fix checkstyle

vigyasharma commented:
The `addIndexes(CodecReaders[])` API is transactional in nature - either all readers get added or all of them fail. Retaining this behavior with concurrent background merges needs some work.

My plan is to defer adding the newly created merged segments to IndexWriter until all merges finish. MergeScheduler threads will do all the work for merge right upto creating segment files, and return. At addIndexes() API, we'll wait for all merges kicked off by the API to complete, and add them to `IndexWriter.segmentInfos` in a sync block only when all of them succeed. For partial failures, we'll delete the interim files created.

Added code for this change to get some feedback on this approach. Tests are failing right now and I'm working on fixing them and adding new ones. Have put `nocommit` comments meanwhile.

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/FieldInfos.java:
```diff
@@ -352,6 +352,14 @@ public FieldDimensions(int dimensionCount, int indexDimensionCount, int dimensio
       this.softDeletesFieldName = softDeletesFieldName;
     }
 
+    public void verifyFieldInfo(FieldInfo fi) {
```
Should this really be public?  Or is it only for `asserting`, in which case maybe make it package private and rename to `boolean assertFieldInfo` or so?

vigyasharma replied:
This method makes sure that incoming index fields are compatible with the destination index, e.g. vectors have the same dimensions and use the same similarity function.

Changed access to package private. I retained method name - `verifyFieldInfo()`, as it internally calls `verifySoftDeletedFieldName` and `verifySameSchema`, which have a similar naming style. 

I can change it if `assert*()` is more in line with the Lucene convention. In which case, I also think that it should only return a condition and the asserts should be done by the caller.. Let me know..

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/IndexWriter.java:
```diff
@@ -3121,147 +3125,265 @@ private void validateMergeReader(CodecReader leaf) {
   public long addIndexes(CodecReader... readers) throws IOException {
     ensureOpen();
-
-    // long so we can detect int overflow:
```
Hmm we lost this comment?  Was that intentional?

vigyasharma replied:
I think it was a miss. Added it back, thanks!

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/IndexWriter.java:
```diff
@@ -3121,147 +3125,265 @@ private void validateMergeReader(CodecReader leaf) {
+      synchronized (this) {
+        ensureOpen();
+        if (merges.areEnabled() == false) {
+          throw new UnsupportedOperationException("Merges are disabled on current writer. " +
```
I think this only happens if IW is aborting/closing?  Maybe throw `AlreadyClosedException` instead?  This message is otherwise sort of confusing to the poor user who sees it.

vigyasharma replied:
Changed to AlreadyClosedException.

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/IndexWriter.java:
```diff
@@ -3121,147 +3125,265 @@ private void validateMergeReader(CodecReader leaf) {
+          mergeScheduler.merge(addIndexesMergeSource, MergeTrigger.ADD_INDEXES);
+          mergesComplete = spec.await();
+        } finally {
+          if (!mergesComplete) {
```
`== false` instead?

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/IndexWriter.java:
```diff
@@ -3121,147 +3125,265 @@ private void validateMergeReader(CodecReader leaf) {
+      }
+
+      MergePolicy mergePolicy = config.getMergePolicy();
+      MergePolicy.MergeSpecification spec = mergePolicy.findMerges(Arrays.asList(readers));
```
Normal merges have the opportunity to "cascade": `MergePolicy` picks a round of initial merges, and as they complete and get swapped in, it is consulted again later and might choose to merge some of those just-merged segments into even bigger ones.  But it looks like we don't allow that here?  We ask once, do those merges, commit to IW's `SegmentInfos`?  Maybe that's fine, it's just different.

vigyasharma replied:
We allow for cascading to proceed in background by calling a `maybeMerge()` at the end of `addIndexes()`.

For `addIndexes(CodecReader...)`, we can only add segments to IW after the readers have gone through `SegmentMerger.merge()`. Until then, we don't have any segments to add, (unlike the `addIndexes(Directory...)` variant, where we already have segments provided in the input). So we wait for the first round of merges to complete, by blocking the API on the merges defined in the MergeSpec.

Once these merges complete, we add segments to IW's SegmentInfos, and invoke a maybeMerge(), which will cascade any merges on top of these segments in

vigyasharma replied:
I've updated java docs for the API, to say that we merge as specified by `MergePolicy#findMerges()`. We are not changing the default behavior right now, and currently, the API will continue to merge all readers into a single segment. I'll update this when we implement [LUCENE-10476](https://issues.apache.org/jira/browse/LUCENE-10476), to add a single segment per reader.

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/IndexWriter.java:
```diff
@@ -3121,147 +3125,265 @@ private void validateMergeReader(CodecReader leaf) {
+        long totalMaxDoc = 0;
+        for (MergePolicy.OneMerge merge: spec.merges) {
+          totalMaxDoc += merge.totalMaxDoc;
+          if (merge.getMergeInfo() != null) {
```
Would this only be `null` if all documents in the merged segment had been deleted?  Maybe add a comment?

vigyasharma replied:
It will be null if the merge never started, for example, say the readers were empty. I'll add a comment.

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/IndexWriter.java:
```diff
@@ -3121,147 +3125,265 @@ private void validateMergeReader(CodecReader leaf) {
+              reserveDocs(totalMaxDoc);
+              success = true;
+            } finally {
+              if (!success) {
```
`== false`?

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/IndexWriter.java:
```diff
@@ -3121,147 +3125,265 @@ private void validateMergeReader(CodecReader leaf) {
+    return seqNo;
+  }
+
+  private class AddIndexesMergeSource implements MergeScheduler.MergeSource {
```
Very cool that we are able to use `MergeSource` for this!!  I think this was added to decouple IW and MS so MS could be better tested (by mocking `MergeSource`s).  Wonderful to see that this better decoupling now helps nice changes like this, too.

vigyasharma replied:
+1, `MergeSource` is a neat abstraction. 
I wanted to move this out of IW, to a different class altogether - try to make IW leaner. But I ran into a bunch of test failures and method access issues that digressed from the main change. Maybe I'll try it again in a separate, follow up PR.

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/MergePolicy.java:
```diff
@@ -567,6 +605,21 @@ public abstract MergeSpecification findMerges(
+   * @param readers set of readers to merge into the main index
+   */
+  public MergeSpecification findMerges(List<CodecReader> readers) throws IOException {
+    OneMerge merge = new OneMerge(readers, leaf -> new MergeReader(leaf, leaf.getLiveDocs()));
```
So this means `MergePolicy` by default matches the behavior `IW.addIndexes(CodecReader[])` today -- doing a single large merge.  Good!

vigyasharma replied:
Yes, this change will modify addIndexes to leverage `MergePolicy` and `MergeScheduler`, without changing its end behavior. 

I am thinking of doing a follow up PR with a merge policy that does a 1:1 addIndexes call, where the merger creates one segment each for every provided reader. It would create a faster, more concurrent `addIndexes(CodecReader...)`, at the cost of deferring some merges to be done later in background. 

Which, I believe is similar to the behavior in `addIndexes(Directory...)` - all incoming segments are simply added to IW, and any merging happens in later in background.

mikemccand replied:
> Which, I believe is similar to the behavior in addIndexes(Directory...) - all incoming segments are simply added to IW, and any merging happens in later in background.

Oooh I really like that!  +1 for follow-on issue here to make both versions behave more consistently.

vigyasharma replied:
Create https://issues.apache.org/jira/browse/LUCENE-10476 for the follow up task.

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/MergePolicy.java:
```diff
@@ -813,12 +866,24 @@ protected final boolean verbose(MergeContext mergeContext) {
 
   static final class MergeReader {
+    final CodecReader codecReader;
     final SegmentReader reader;
```
Hmm it's a little spooky/confusing/smelly having both of these -- does anything still require accessing via `SegmentReader`?  Can they switch to `CodecReader`?

vigyasharma replied:
So far, I see `SegmentReader` being used only in `ReadersAndUpdates` [here](https://github.com/apache/lucene/blob/main/lucene%2Fcore%2Fsrc%2Fjava%2Forg%2Fapache%2Flucene%2Findex%2FReadersAndUpdates.java#L786-L793). I am not really sure if `createNewReaderWithLatestLiveDocs()` can also work with a CodecReader.. Will look into this more.

vigyasharma replied:
I looked into this more, and I'm not able to get rid of either of them in the current structure. SegmentReader seems to hold important information about segments that is required by consumers in different places, like the regular segment merges, so I can't always just use CodecReader in OneMerge objects.

AddIndexes() can only work with CodecReaders, as there are no segments to begin with (which means I can also not create a segment reader from the provided codec readers here).

Maybe, I can extend `OneMerge` to create a `ReaderOneMerge` class that only works with CodecReaders? Would love to hear your thoughts, or any other better ideas..

mikemccand replied:
OK let's leave it as is for now -- we can improve that later?

mikemccand commented on lucene/core/src/java/org/apache/lucene/index/NoMergePolicy.java:
```diff
@@ -39,6 +40,11 @@ public MergeSpecification findMerges(
 
+  @Override
+  public MergeSpecification findMerges(List<CodecReader> readers) throws IOException {
+    return null;
```
Hmm should `null` even be allowed by this new API?  Won't this throw `NPE` if someone tries to `addIndexes`?

vigyasharma replied:
In my current impl, addIndexes simply completes without adding anything if the API returns null. I now realize, this is misleading for the user. Callers of addIndexes may not know about the configured merge policy. I should probably throw an exception if I get a null here?
On similar lines, what should we do if the API call had some CodecReaders, but findMerges returns an empty spec.? Should we error out? (Will add a log for this either way.)

Re: letting the API return null, I see that other variants of `findMerges` in this policy, all return `null`, and all callers of findMerges seem to be doing the null checks. Personally, I prefer disallowing the null return values, but I wanted to keep this in sync with the other variants.

vigyasharma replied:
There are consumers of `NoMergePolicy` that expect `addIndexes(CodecReaders...)` to work, like `IndexRearranger`. My understanding is that NoMergePolicy is meant to turn off any merges on existing segments. Since this API uses merges to add new readers to the index, I've changed it here to return the default base class impl.

It's a little off because the name is NoMergePolicy, but I think(hope) it aligns with the intent behind that class. I've added some comments to call this out. Let me know if we should instead create a different MergePolicy that only allows fineMerges(readers).. (it would be exactly like NoMergePolicy except for one api).

mikemccand reviewed (commented):
I really love this approach, of letting `MergePolicy` guide exactly how merging of `addIndexes(CodecReader[])` will happen.

What remains to bring this out of WIP?  I see a few `// nocommits` still to fix ... can you post the failing tests details here maybe?  This is certainly tricky code; thank you for tackling it!

vigyasharma committed:
PR comments

vigyasharma commented:
Thanks for reviewing this @mikemccand ! I've addressed the ones that had minor changes, and am working on the larger ones.

> What remains to bring this out of WIP?

There are 2 existing tests that are breaking:
```java
- org.apache.lucene.index.TestAddIndexes.testAddIndicesWithSoftDeletes (:lucene:core)
- org.apache.lucene.index.TestIndexSorting.testAddIndexesWithDeletions (:lucene:core)
```

I've been looking at `testAddIndicesWithSoftDeletes`. IW does a `rollbackInternal()` on shutDown, where it is failing on this assert - `assert pendingNumDocs.get() == segmentInfos.totalMaxDoc()`. I'm guessing my code is reserving more docs than it should.. is that the right direction? 
Here's a longer error stack trace - 
```java
java.lang.AssertionError: pendingNumDocs 7 != 5 totalMaxDoc
   at __randomizedtesting.SeedInfo.seed([63C0554849BFDEC5:4F94AA0CFCBAFADD]:0)
   at org.apache.lucene.index.IndexWriter.rollbackInternal(IndexWriter.java:2428)
   at org.apache.lucene.index.IndexWriter.shutdown(IndexWriter.java:1334)
   at org.apache.lucene.index.IndexWriter.close(IndexWriter.java:1362)
   at org.apache.lucene.util.IOUtils.close(IOUtils.java:85)
   at org.apache.lucene.util.IOUtils.close(IOUtils.java:72)
   at org.apache.lucene.index.TestAddIndexes.testAddIndicesWithSoftDeletes(TestAddIndexes.java:1521)
```
 
Outside of fixing existing tests, I want to add some new tests, especially around cases where one of the merges fails, causing the whole API to fail (i.e. test that transactionality was retained).

vigyasharma committed:
Fix pendingDocs - numDocs mismatch bug

vigyasharma committed:
Tests with 1-1 merges and partial merge failures

vigyasharma committed:
variable renaming and better comments

vigyasharma committed:
add test for partial merge failures. change tests to use 1-1 findmerges

vigyasharma committed:
abort pending merges gracefully

vigyasharma committed:
test null and empty merge specs

vigyasharma committed:
test interim files are deleted

vigyasharma committed:
test with empty readers

vigyasharma committed:
test cascading merges triggered

vigyasharma committed:
remove nocommits

vigyasharma committed:
gradle check errors

vigyasharma committed:
remove unused line

vigyasharma committed:
remove printf

vigyasharma committed:
spotless apply

vigyasharma committed:
update TestIndexWriterOnDiskFull to accept mergeException from failing addIndexes calls

vigyasharma committed:
return singleton reader mergespec in NoMergePolicy

vigyasharma commented:
### New Changes:
 1. `addIndexes(CodecReader...)` merges are abortable now. 
    * Earlier, `addIndexes(CodecReader...)` triggered merges directly via SegmentMerger. As a result, the running merges could be tracked in `Set<SegmentMerger> runningAddIndexesMerges`, but any pending merge operations could not be aborted preemptively.
    * With this change, merges are defined via `MergePolicy.OneMerge` objects, and scheduled by the `MergeScheduler`. These are aborted when IndexWriter is rolled back or closed.
    * removed corresponding TODO - `“TODO: somehow we should fix this merge so it’s abortable so that IW.close(false) is able to stop it”`
2. Updated NoMergePolicy to return the default MergeSpec for `findMerges(List<CodecReader> readers)`, as it is required by consumers of addIndexes that do use this policy to avoid background segment merges, like `IndexRearranger`.

### Tests Added:
1. Test addIndexes with a MergePolicy that creates one merge per CodecReader.
2. Test transactionality when only a subset of triggered merges succeed.
3. Test interim files are deleted when a partial subset of merges fails, causing the API to fail.
4. Tests for null and empty MergeSpec returned from MergePolicy.
5. Test no merges are triggered if empty readers provided.
6. Test cascading merges get triggered after each `addIndexes(CodecReader...)` call.

### Existing Tests Expanded:
1. Added a test MergePolicy that creates one merge per CodecReader. This is used with multiple existing concurrency tests, that check for steady state calls, concurrent operations overlapping with addIndexes, IW closed (with and without wait) with a conc. running addIndexes, and IW rollback. 
2. Tested that pending merges get aborted when IW is rolled back.

vigyasharma commented:
I think this PR has reasonable test coverage and can come out of WIP now. Will continue to iterate on the PR comments and suggestions..

vigyasharma commented:
`./gradlew check` had passed on my local machine. However, some tests have randomization and don't always invoke `addIndexes(CodecReaders...)`, and have failed in the run above.

Is there a way to ensure all code paths in the random tests get executed? I want to run the tests that invoke `addIndexes(CodecReader...)` for some random flag value.

vigyasharma committed:
rethrow exceptions seen in merge threads on failure

vigyasharma committed:
spotless apply

vigyasharma committed:
update test to new exception type thrown

vigyasharma committed:
spotlessApply

vigyasharma committed:
test for maxDoc limit in IndexWriter

vigyasharma committed:
spotlessApply

vigyasharma commented:
1. Found a better way to handle addIndexes failures due to background merges -
OneMerge objects store the error they run into, which can be pulled via `OneMerge.getException()`. We `rethrow` these errors if they turn out to be the cause for addIndexes failure. This fixes tests which expect certain exceptions, like random IO exceptions thrown by MockDirectoryWrapper in `BasePointsFormatTestCase#testWithExceptions`.

2. Added a test for validate we API and destination writer state when we hit `IndexWriter.actualMaxDocs` limit. 
3. Fixed breaking tests from the last run.

mikemccand commented:
> Is there a way to ensure all code paths in the random tests get executed? I want to run the tests that invoke `addIndexes(CodecReader...)` for some random flag value.

Alas, I don't think we have a way to ensure ALL paths get executed -- that would likely be combinatorically prohibitive?  But sure would be nice to have some control over this.

But we do have `gradle beast`, which will run the same test case over and over with different random seeds and hopefully over time provoke any latent failures that one or two successful runs failed to uncover?

vigyasharma commented:
Thanks @mikemccand . I figured that I had changed the exception signature for addIndexes() in my last revision. I was throwing a blanket MergePolicyException when addIndexes failed to merge all the readers in; and a lot of these tests were designed to look for specific exceptions that would've earlier gotten thrown from the SegmentMerger code  running in main addIndexes thread.. This was causing the randomly invoked tests to fail.

I changed that, to `rethrow` the exceptions seen by any failed background OneMerges, and that appeased the gradle gods and builds began to pass again. 

I think this is better, because now the API throws the actual error seen by a thread doing the merge, instead of some blanket failure message. In general, I have to say, I'm fascinated by the breadth of test cases around this API, and the code framework present to add more tests.

mikemccand commented:
> This method makes sure that incoming index fields are compatible with the destination index, e.g. vectors have the same dimensions and use the same similarity function.
> 
> Changed access to package private. I retained method name - verifyFieldInfo(), as it internally calls verifySoftDeletedFieldName and verifySameSchema, which have a similar naming style.
> 
> I can change it if assert*() is more in line with the Lucene convention. In which case, I also think that it should only return a condition and the asserts should be done by the caller.. Let me know..

OK let's leave it as `verify` now -- when in Rome!

mikemccand reviewed (approved):
Thanks @vigyasharma -- this is a super exciting change!  It generalizes `addIndexes` to use `MergePolicy/Scheduler` so that incoming indices can be added concurrently.  I think it's ready, but needs rebasing to resolve conflicts?

vigyasharma committed:
Use DocValuesIterator instead of DocValuesFieldExistsQuery for counting soft deletes

vigyasharma committed:
spotless apply

vigyasharma committed:
change exception message for closed IW

vigyasharma committed:
remove non-essential comments

vigyasharma committed:
update api doc string

vigyasharma committed:
doc string update

vigyasharma committed:
spotless

vigyasharma commented on lucene/core/src/java/org/apache/lucene/index/IndexWriter.java:
```diff
@@ -3127,143 +3135,307 @@ public long addIndexes(CodecReader... readers) throws IOException {
+        ensureOpen();
+        if (merges.areEnabled() == false) {
+          throw new AlreadyClosedException(
+              "Merges are disabled on current writer. "
```
Good point, updated the exception message

vigyasharma commented:
Thanks for reviewing these rather lengthy changes, @mikemccand! I've rebased this on to the latest main branch, and updated doc string and exception message.

vigyasharma committed:
Changes file entry

vigyasharma commented:
The build failure is on `TestTaxonomyFacetAssociations.testFloatAssociationRandom`, and I'm unable to repro it without the randomization seed. 

The failure (with randomization seed) is because of a mismatch in (SUM) aggregated multi-valued, `float_random` facet field. We accept an [error delta](https://github.com/apache/lucene/blob/main/lucene/facet/src/test/org/apache/lucene/facet/taxonomy/TestTaxonomyFacetAssociations.java#L445) of 1 in this aggregation, but for the failing random seed, the delta is 1.3. For the failing random seed, the test also fails on the `main` branch.

I'm not sure if this is related to the PR's changes (I don't see any related APIs being invoked by the test). My hunch is that it is some floating point approximation error.  
@gsmiller,  i see you worked on it recently, what are your thoughts? Would it make sense increase the `delta` in assert to 1.5?

--

**Failing Test**
```ruby
./gradlew test --tests TestTaxonomyFacetAssociations.testFloatAssociationRandom \ 
-Dtests.seed=4DFBA8209AC82EB2 -Dtests.slow=true -Dtests.locale=fr-VU \
-Dtests.timezone=Europe/Athens -Dtests.asserts=true -Dtests.file.encoding=UTF-8
```
**Failure:**
```ruby
org.apache.lucene.facet.taxonomy.TestTaxonomyFacetAssociations > testFloatAssociationRandom FAILED
    java.lang.AssertionError: expected:<2409060.8> but was:<2409059.5>
        at __randomizedtesting.SeedInfo.seed([4DFBA8209AC82EB2:1EC25C9148A9BF04]:0)
        at junit@4.13.1/org.junit.Assert.fail(Assert.java:89)
        at junit@4.13.1/org.junit.Assert.failNotEquals(Assert.java:835)
        at junit@4.13.1/org.junit.Assert.assertEquals(Assert.java:577)
        at junit@4.13.1/org.junit.Assert.assertEquals(Assert.java:701)
        at org.apache.lucene.facet.taxonomy.TestTaxonomyFacetAssociations.validateFloats(TestTaxonomyFacetAssociations.java:445)
        at org.apache.lucene.facet.taxonomy.TestTaxonomyFacetAssociations.testFloatAssociationRandom(TestTaxonomyFacetAssociations.java:256)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
       ...
```

**Passing Test:**
```ruby
./gradlew test --tests TestTaxonomyFacetAssociations.testFloatAssociationRandom \
-Dtests.slow=true -Dtests.locale=fr-VU -Dtests.timezone=Europe/Athens \
-Dtests.asserts=true -Dtests.file.encoding=UTF-8
```

mikemccand commented:
Thanks for digging into that failure @vigyasharma.

> For the failing random seed, the test also fails on the main branch.

OK, so this is pre-existing!  Could you open a separate spinoff issue and post the failure details there?  It should not block this awesome change.

mikemccand commented:
I restarted the `gradle test` job.

mikemccand commented on lucene/CHANGES.txt:
```diff
@@ -89,6 +89,9 @@ New Features
 Improvements
 ---------------------
 
+* LUCENE-10216: Use MergePolicy to define and MergeScheduler to trigger the reader merges
```
Thanks for adding `CHANGES` entry, but we usually append to the end of the section not the start :)

mikemccand commented on lucene/core/src/test/org/apache/lucene/index/TestAddIndexes.java:
```diff
@@ -1014,8 +1304,6 @@ public void testAddIndexesWithRollback() throws Throwable {
 
     c.closeDir();
-
-    assertTrue(c.failures.size() == 0);
```
Hmm why did we need to remove this assertion?

vigyasharma replied:
We don't. This checks for failures when writer.close() was called. We are able to abort merges and don't have any failures, so this assertion is valid. I think I removed it while debugging some breaking tests and missed adding it back. Added it back now. Thanks!

mikemccand commented on lucene/core/src/test/org/apache/lucene/index/TestIndexWriterOnDiskFull.java:
```diff
@@ -352,15 +352,15 @@ public void testAddIndexOnDiskFull() throws IOException {
             }
 
-          } catch (IllegalStateException | IOException e) {
+          } catch (IllegalStateException | IOException | MergePolicy.MergeException e) {
```
Hmm can you give an example exception when we now throw `MergePolicy.MergeException` inside the `try`?

vigyasharma replied:
If the merge threads triggered for addIndexes readers fail, they throw a merge exception, which is silently recorded in that OneMerge object. After all threads join, we check if merge passed on all threads, and if not, we rethrow this MergeException, to surface the actual cause of API failure. This is [done here](https://github.com/apache/lucene/pull/633/files/826116d47268714685ad42988ae5f9b66bac6673#diff-edbb11369e2e565210ea9761ee167ac0c6a39521c0e71f8af24f25f4092c740aR3224-R3229). If all threads passed, we proceed to register the new segmentinfo.

`testAddIndexOnDiskFull()` simulates this by randomly filling up the disk and causing these merge threads to fail.

mikemccand reviewed (commented):
Hi @vigyasharma -- I had a couple questions about some changes to existing unit tests.

I think it would be worth implementing the new `MergePolicy` method in either `MockRandomMergePolicy` or `AlcoholicMergePolicy` or maybe both to exercise concurrent `addIndexes` across any tests today or future tests that use `addIndexes`?

vigyasharma committed:
simplify findMerges API, add 1-1 merges to MockRandomMergePolicy

vigyasharma committed:
update merge policies to new api

vigyasharma committed:
remove unused imports

vigyasharma committed:
spotless apply

vigyasharma committed:
move changes entry to end of list

vigyasharma committed:
fix testAddIndicesWithSoftDeletes

vigyasharma committed:
test with 1-1 merge policy always enabled

vigyasharma committed:
please spotcheck

vigyasharma committed:
tidy

vigyasharma committed:
test - never use 1-1 merge policy

vigyasharma committed:
use 1-1 merge policy randomly

vigyasharma commented:
Updated the `MockRandomMergePolicy` to randomly use the more concurrent version of `findMerges(CodecReaders[])` with a 50% probability. This uncovered some failing tests that were asserting directly on segments within a writer, and expected a single segment out of the API. Fixed broken tests. 

Kudos to Lucene's randomized testing for catching such edge cases!

rmuir commented on lucene/test-framework/src/java/org/apache/lucene/tests/index/MockRandomMergePolicy.java:
```diff
@@ -86,6 +86,20 @@ public MergeSpecification findMerges(
+    if (random.nextBoolean() == true) {
+      return super.findMerges(readers);
+    } else {
+      // create a oneMerge for each reader to let them get concurrently processed by addIndexes()
```
Can we do this much much less often, possibly only at nightly, or not at all?

It is very easy to transition from "repeatable randomized tests" to "chaotic flaky non-reproducing madness" by adding a bunch of concurrency everywhere. It's not the way.

uschindler replied:
Yes please don't do this for all tests. Not even nightly.

Please add a specific test only for concurrently merging addindexes, bit not make it the default for all uses of IndexWriter in unit tests.

msokolov replied:
Perhaps create specific tests for the ones that actually failed when this was enabled for all tests?

uschindler replied:
I don't think those tests are wrong. The tests are there to test specific merging behaviour. If they only expect one segment they should only get one segment. Those tests should not suddenly get a concurrent addIndexes behaviour. That's not what they assert. So there is definitely nothing wrong with those tests. Changes in them should be reverted!

uschindler replied:
But there should be an additional test that checks if concurrent addIndexes works as expected.

vigyasharma replied:
Thanks for reviewing this @rmuir, @uschindler. I understand your concern. Failures on concurrent code are already tough to reproduce; it gets much harder to do this from a randomized test seed. 

It is worth noting that the `findMerges(CodecReader[])` API modified in `MockRandomMergePolicy`, is only invoked within the `addIndexes(CodecReaders[])` API, which should contain its scope of impact. But I definitely don't want to add noise with random failures on distantly related tests. I'll remove this new policy from `MockRandomMergePolicy`, which is frequently used for `IndexWriterConfig` in multiple tests.

For my awareness, what is the best practice in Lucene for testing non-default, concurrency related code paths? Randomized tests help validate infrequently triggered code, but concurrency poses has reproducibility challenges. For such cases, do we try to rely entirely on change-specific concurrency unit tests?

vigyasharma replied:
> Please add a specific test only for concurrently merging addIndexes, but not make it the default for all uses of IndexWriter in unit tests.

I have specific tests for this change - to test concurrent execution [[1]](https://github.com/apache/lucene/pull/633/files#diff-ce6b91c31190c4956eee8a734550af25047d33b92a71c3d7303eac4cc408c504R741), rollbacks and transactional behavior [[2]](https://github.com/apache/lucene/pull/633/files#diff-ce6b91c31190c4956eee8a734550af25047d33b92a71c3d7303eac4cc408c504R790), boundary conditions [[3]](https://github.com/apache/lucene/pull/633/files#diff-ce6b91c31190c4956eee8a734550af25047d33b92a71c3d7303eac4cc408c504R820), [[4]](https://github.com/apache/lucene/pull/633/files#diff-ce6b91c31190c4956eee8a734550af25047d33b92a71c3d7303eac4cc408c504R838), [[5]](https://github.com/apache/lucene/pull/633/files#diff-ce6b91c31190c4956eee8a734550af25047d33b92a71c3d7303eac4cc408c504R882), [[6]](https://github.com/apache/lucene/pull/633/files#diff-ce6b91c31190c4956eee8a734550af25047d33b92a71c3d7303eac4cc408c504R926),  triggering background merges [[7]](https://github.com/apache/lucene/pull/633/files#diff-ce6b91c31190c4956eee8a734550af25047d33b92a71c3d7303eac4cc408c504R917), and changes in concurrency tests that check for IW abort, rollback and competing concurrent operations [[8]](https://github.com/apache/lucene/pull/633/files#diff-ce6b91c31190c4956eee8a734550af25047d33b92a71c3d7303eac4cc408c504R993-R994). 

Would love to get your inputs on any other scenarios that I should test.

vigyasharma replied:
> Perhaps create specific tests for the ones that actually failed when this was enabled for all tests?

> ...So there is definitely nothing wrong with those tests. Changes in them should be reverted!


The only test that failed was `testAddIndicesWithSoftDeletes()`. It tests two different scenarios of soft deletes. One with `addIndexes(readers)` where it validates the no. of soft deletes after adding readers to an empty index. 
The other is with `addIndexes(wrappedReaders)` where `wrappedReaders` filter out the soft deletes, and so the resulting writer has `maxDoc == wrappedReader.numDocs()`.

The test was using `writer.cloneSegmentInfos().info(0)` to count number of soft deletes in resulting index for scenario-1. This requires the index to be empty and for only a single segment to get added.  I believe new segments get added at the end of the list. If the intent here is to ensure that only one segment is added, maybe we should make it something like `writer.cloneSegmentInfos().info(numSegments - 1)`? Or separately assert on the no. of new segments added? It does seem outside the purpose of this test, which is to check for soft delete retention.

I [changed](https://github.com/apache/lucene/pull/633/files#diff-ce6b91c31190c4956eee8a734550af25047d33b92a71c3d7303eac4cc408c504R1796-R1800) it to use a sum of softDeleteCount from every segment, which, per my understanding, is also what [`writer.getDocStats()`](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/index/IndexWriter.java#L6192) does. The test passes with both concurrent and single merge addIndexes, with this change. Let me know if I should revert this to its former state.

uschindler replied:
This single test looks still fine. You talked of multiple tests, so I was unsure which ones were changed because of your statement.

vigyasharma replied:
Ah, my bad. I'll update that comment. Thanks.

mikemccand replied:
Woops, sorry @vigyasharma: it was my suggestion to try to use the concurrent `addIndexes` sometimes, randomly, via `MockMergePolicy` or `AlcoholicMergePolicy` in the few tests that also test `addIndexes(CodecReader[])`!!!

Apparently that was controversial -- I love it :)  So let's just revert that and rely on the awesome dedicated unit tests @vigyasharma already added?

Remember that this feature is not enabled by default -- users who upgrade will still see the slow single-threaded `addIndexes`, unless they customize their `MergePolicy`.  If we make this PR/change we can iterate over time as users try to adopt it and perhaps eventually make it the default.  Concurrent background merging was also controversial when it was first proposed, yet concurrency is vital to take advantage of modern hardware.  Lucene shows its age by having so many single threaded operations.

Also, it's unfortunate that improving Lucene's concurrency risks non-reproducible tests.  I wish the JVM would enable a seed-reproducible mode where it consistently scheduled threads down to the byte-code or ASM instruction (I know this is hard, but, still).  This problem will only get worse for us as we try to improve Lucene's default ability to leverage all the cores in modern CPUs.

rmuir replied:
> Thanks for reviewing this @rmuir, @uschindler. I understand your concern. Failures on concurrent code are already tough to reproduce; it gets much harder to do this from a randomized test seed.
> 
> It is worth noting that the `findMerges(CodecReader[])` API modified in `MockRandomMergePolicy`, is only invoked within the `addIndexes(CodecReaders[])` API, which should contain its scope of impact. But I definitely don't want to add noise with random failures on distantly related tests. I'll remove this new policy from `MockRandomMergePolicy`, which is frequently used for `IndexWriterConfig` in multiple tests.
> 

This is true, but also part of the risk: `addIndexes(CodecReaders[])` is already pretty scary, correctness-wise. This is typically used to expertly "transform" an index, otherwise you wouldn't use this API. It is probably undertested today, I just don't want to mix concurrency randomly into the few tests it has.

> For my awareness, what is the best practice in Lucene for testing non-default, concurrency related code paths? Randomized tests help validate infrequently triggered code, but concurrency poses has reproducibility challenges. For such cases, do we try to rely entirely on change-specific concurrency unit tests?

Well, we need to think about it. I actually hate that ConcurrentMergeScheduler gets used the way it does in the tests, for this same reason. And that's the actual problem here... I think, not your logic in the mergepolicy? So really my concerns are around mergescheduler in tests preventing reproducibility. CMS is the default if someone tests with just `new IndexWriterConfig` and it is almost always the default if they use `LuceneTestCase.newIndexWriterConfig`.

Maybe we can improve our tests to use `SerialMergeScheduler` and `ConcurrentMergeScheduler(n=1)` more often, so that reproducibility is improved across the board. We can have separate specific tests using `ConcurrentMergeScheduler(n>1)`? 

It really isn't anything new from your change here, I think i was just a bit frustrated the last time i tried to reproduce a seed recently (https://issues.apache.org/jira/browse/LUCENE-10529), so I want us to be careful about blanket application of concurrency to a ton of existing tests.

vigyasharma replied:
> Well, we need to think about it. I actually hate that ConcurrentMergeScheduler gets used the way it does in the tests, for this same reason. And that's the actual problem here... I think, not your logic in the mergepolicy?

Yes, the concurrency here comes from concurrent merge scheduler. If this policy was to run with the `SerialMergeScheduler`, there would be no concurrency. 


Is it that randomized tests (without concurrency) cannot always be reproduced with their random seed (like [LUCENE-10529](https://issues.apache.org/jira/browse/LUCENE-10529)? And so, with mixing randomized tests and concurrency, we blend two unknowns. We are not sure if we're missing the params with which the test failed, or if the random seed worked but it's a race condition issue..


> Maybe we can improve our tests to use `SerialMergeScheduler` and `ConcurrentMergeScheduler(n=1)` more often, so that reproducibility is improved across the board. We can have separate specific tests using `ConcurrentMergeScheduler(n>1)`?

We should probably do that for tests that are validating some non-concurrent logic, and not asserting any CMS specific behavior. On similar lines, what do you think about moving concurrency tests for an area, into a separate class/test suite - like `TestAddIndexesConcurrency` for `addIndexes`?. Then, for changes that touch concurrency, contributors can separately run those test suites on repeat to get better confidence. It would also be easier to setup such classes for concurrency (like using n threads in CMS), and avoid adding any randomness to it.

rmuir replied:
> Is it that randomized tests (without concurrency) cannot always be reproduced with their random seed (like [LUCENE-10529](https://issues.apache.org/jira/browse/LUCENE-10529)? And so, with mixing randomized tests and concurrency, we blend two unknowns. We are not sure if we're missing the params with which the test failed, or if the random seed worked but it's a race condition issue..

you got it.



> We should probably do that for tests that are validating some non-concurrent logic, and not asserting any CMS specific behavior. On similar lines, what do you think about moving concurrency tests for an area, into a separate class/test suite - like `TestAddIndexesConcurrency` for `addIndexes`?. Then, for changes that touch concurrency, contributors can separately run those test suites on repeat to get better confidence. It would also be easier to setup such classes for concurrency (like using n threads in CMS), and avoid adding any randomness to it.

I really like this idea of just introducing a "convention" for test naming so that when such a test fails (e.g. in CI), it is very clear that the issue may be concurrency related! And making it easy for ppl with lots of cpus to "beast" concurrency tests.

vigyasharma replied:
Great! I can pilot this in a follow up PR for `TestAddIndexes`, and we could incrementally move pieces as we touch them. Should we start a Jira to discuss the naming convention? How about something like `testConcurrent*` e.g. `testConcurrentAddIndexesWithRollback`?

rmuir replied:
+1 to make a followup issue. Could use some more discussion and ideas. Yet another idea is to add a test group (e.g. `@Concurrent`) to make it easy to run tests in that group: https://github.com/apache/lucene/blob/main/help/tests.txt#L67-L81

vigyasharma replied:
Created https://issues.apache.org/jira/browse/LUCENE-10566

vigyasharma committed:
Remove concurrent addIndexes findMerges from MockRandomMergePolicy

vigyasharma committed:
Bug Fix: RuntimeException in addIndexes

Aborted pending merges were slipping through the merge exception check in
API, and getting caught later in the RuntimeException check.

vigyasharma commented:
Removed the randomly returned high concurrency mergeSpec from `findMerges(CodecReaders...)` in `MockRandomMergePolicy`. Added a check to gracefully catch pending merges that are aborted before they start, due to a closing `IndexWriter`. Such merges do not see any exceptions (since they never started). We now catch them and fail the API with `MergePolicyException` (aborted merges) instead of a less descriptive `RuntimeException`.

Ran the `TestAddIndexes` suite multiple times on repeat to catch concurrency issues.
```python
% ./gradlew :lucene:core:test --tests "org.apache.lucene.index.TestAddIndexes" -Dtests.iters=500 
...
:lucene:core:test (SUCCESS): 17500 test(s)
The slowest tests (exceeding 500 ms) during this run:
   2.77s TestAddIndexes.testCascadingMergesTriggered {seed=[995488156EAA9213:C061B8FC39526277]} (:lucene:core)
   2.57s TestAddIndexes.testCascadingMergesTriggered {seed=[995488156EAA9213:6449415B81ED55E8]} (:lucene:core)
   2.50s TestAddIndexes.testAddIndexesWithConcurrentMerges {seed=[995488156EAA9213:8DDA90E4057EEECC]} (:lucene:core)
   2.34s TestAddIndexes.testCascadingMergesTriggered {seed=[995488156EAA9213:98ABD78A2EF738F8]} (:lucene:core)
   1.86s TestAddIndexes.testCascadingMergesTriggered {seed=[995488156EAA9213:F4015EAB7FEBE4C4]} (:lucene:core)
   1.83s TestAddIndexes.testCascadingMergesTriggered {seed=[995488156EAA9213:4C70D3A5B0897539]} (:lucene:core)
   1.65s TestAddIndexes.testAddIndexesWithThreads {seed=[995488156EAA9213:9168C57ACC52DE66]} (:lucene:core)
   1.62s TestAddIndexes.testAddIndexesWithThreads {seed=[995488156EAA9213:777C792E198970FF]} (:lucene:core)
   1.53s TestAddIndexes.testAddIndexesWithRollback {seed=[995488156EAA9213:FC3F09AFA640AEA1]} (:lucene:core)
   1.53s TestAddIndexes.testAddIndexesWithRollback {seed=[995488156EAA9213:CDD2AABCE4428929]} (:lucene:core)
The slowest suites (exceeding 1s) during this run:
  1347.53s TestAddIndexes (:lucene:core)

BUILD SUCCESSFUL in 23m 4s
```

vigyasharma committed:
tidy

mikemccand commented:
Thanks @vigyasharma -- I'll review again soon.

vigyasharma committed:
Rebase on main. Move changes to 10.0

vigyasharma commented:
Rebased on main. Moved `changes` entry to `10.0.0` since we're close to cutting `9.2.0`, and it would be good to bake this in CI builds for a while.

mikemccand commented:
I think this is ready!

What a great change, giving `MergePolicy` purview over how `addIndexes` does its merging too.  I'm also beasting `TestAddIndexes` and will report back.

I agree we should push only to `main` to start and let this bake for a while before backporting to 9.x.

Thank you @vigyasharma!

mikemccand commented:
Hmm I beasted the PR on current `main` branch using `./gradlew :lucene:core:test --tests "org.apache.lucene.index.TestAddIndexes" -Dtests.iters=500 `.

The first time, it seemed to deadlock, but I screwed up trying to get the thread stacks.

The 2nd time it failed with this exciting failure!  It looks like we are missing some locking on updating IW's `HashMap` tracking merges?:

```
org.apache.lucene.index.TestAddIndexes > test suite's output saved to /l/vigya/lucene/core/build/test-results/test/outputs/OUTPUT-org.apache.lucene.index.TestAddIndexes.txt, copied below:
  1> java.util.ConcurrentModificationException
  1>    at java.base/java.util.HashMap$HashIterator.nextNode(HashMap.java:1597)
  1>    at java.base/java.util.HashMap$KeyIterator.next(HashMap.java:1620)
  1>    at org.apache.lucene.index.IndexWriter.forceMerge(IndexWriter.java:2122)
  1>    at org.apache.lucene.index.IndexWriter.forceMerge(IndexWriter.java:2081)
  1>    at org.apache.lucene.index.TestAddIndexes$CommitAndAddIndexes3.doBody(TestAddIndexes.java:1221)
  1>    at org.apache.lucene.index.TestAddIndexes$RunAddIndexesThreads$1.run(TestAddIndexes.java:1023)
   >     java.lang.AssertionError
   >         at __randomizedtesting.SeedInfo.seed([3AB557F4BFC8763A:D625EE2D285D2CAD]:0)
   >         at org.junit.Assert.fail(Assert.java:87)
   >         at org.junit.Assert.assertTrue(Assert.java:42)
   >         at org.junit.Assert.assertTrue(Assert.java:53)
   >         at org.apache.lucene.index.TestAddIndexes.testAddIndexesWithRollback(TestAddIndexes.java:1305)
   >         at jdk.internal.reflect.GeneratedMethodAccessor38.invoke(Unknown Source)
   >         at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
   >         at java.base/java.lang.reflect.Method.invoke(Method.java:568)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1754)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:942)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:978)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:992)
   >         at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
   >         at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
   >         at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
   >         at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
   >         at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
   >         at org.junit.rules.RunRules.evaluate(RunRules.java:20)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:370)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:819)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:470)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:951)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:836)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:887)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:898)
   >         at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
   >         at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
   >         at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
   >         at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
   >         at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
   >         at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
   >         at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
   >         at org.junit.rules.RunRules.evaluate(RunRules.java:20)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:370)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:826)
   >         at java.base/java.lang.Thread.run(Thread.java:833)
  2> NOTE: reproduce with: gradlew test --tests TestAddIndexes.testAddIndexesWithRollback -Dtests.seed=3AB557F4BFC8763A -Dtests.locale=ru-KZ -Dtests.timezone=Europe/Vilnius -Dtests.asserts=true -Dtests\
.file.encoding=UTF-8
  2> NOTE: leaving temporary files on disk at: /l/vigya/lucene/core/build/tmp/tests-tmp/lucene.index.TestAddIndexes_3AB557F4BFC8763A-001
  2> NOTE: test params are: codec=Asserting(Lucene92), sim=Asserting(RandomSimilarity(queryNorm=false): {foo=DFR I(ne)L1, id=DFR I(n)LZ(0.3), content=F2LOG}), locale=ru-KZ, timezone=Europe/Vilnius
  2> NOTE: Linux 5.17.5-arch1-1 amd64/Oracle Corporation 17.0.1 (64-bit)/cpus=128,threads=1,free=97508840,total=536870912
  2> NOTE: All tests run in this JVM: [TestAddIndexes]
```

vigyasharma commented:
@mikemccand Thanks for beasting it and uncovering this issue! 
I tried to do the same on my box but haven't had any luck so far. Any ideas on how I can make it more likely to happen? How many cores (and threads) did you run this on?

Side Note: it might be helpful to have a gradle parameter for thread count for concurrency tests. Maybe something that goes with the [Concurrent](https://issues.apache.org/jira/browse/LUCENE-10566) test group? I'm not sure if there's one for this already (the test seems hardcoded to `int numThreads = TEST_NIGHTLY ? 5 : 2;`.

I tried reading through code to figure this out. I suspect this is because the `synchronized AddIndexesMergeSource::getNextMerge()` function is really locking on the inner class object, and not on IndexWriter. This works fine for `pendingAddIndexesMerges` because that is an inner class object, but probably hits race conditions for `runningMerges` which belongs to `IndexWriter`.

I could either wrap the runningMerges update with a `synchronized (IndexWriter.this) {}`, or make `runningMerges` a `synchronizedSet`. I like the second approach as it automatically fixes this at all other places. What do you think?

vigyasharma committed:
Synchronize IW.AddIndexesMergeSource on outer class IW object

vigyasharma committed:
tidy

vigyasharma commented:
> I could either wrap the runningMerges update with a synchronized (IndexWriter.this) {}, or make runningMerges a synchronizedSet. I like the second approach as it automatically fixes this at all other places.

I decided to go with the first approach of wrapping critical sections with a `synchronized (IndexWriter.this) {}`. This feels simpler to reason about as we're synchronizing on a single object - `IndexWriter.this`. 

A `runningMerges` synchronizedSet, would've created three different objects that were getting locked at different places - the IndexWriter object, the synchronizedSet lock, and the AddIndexesMergeSource, which has sync functions so that running and pending merge queues can be transactionally updated. Reasoning about, and avoiding deadlocks is simpler with a single IW lock. And given the nature of these critical sections, I don't think this affects concurrency by much.

Also rebased on the latest main.
